mitiq/rem/rem.py

```diff
@@ -48,66 +49,86 @@ def execute_with_rem(
         The expectation value estimated with REM.
     """
     if not isinstance(executor, Executor):
-        executor = Executor(executor)
-
-    result = executor._run([circuit])
-    noisy_result = result[0]
-    if not isinstance(noisy_result, MeasurementResult):
-        raise TypeError("Results are not of type MeasurementResult.")
+        executor_obj = Executor(executor)
 
-    mitigated_result = mitigate_measurements(
-        noisy_result, inverse_confusion_matrix
+    executor_with_rem = mitigate_executor(
+        executor_obj, inverse_confusion_matrix=inverse_confusion_matrix
     )
 
-    return observable._expectation_from_measurements([mitigated_result])
+    # Since the input is an Executor the output is an Executor
+    executor_with_rem = cast(Executor, executor_with_rem)
+
+    return executor_with_rem.evaluate(circuit, observable)[0]
 
 
 def mitigate_executor(
-    executor: Callable[[QPROGRAM], MeasurementResult],
-    observable: Observable,
+    executor: Union[Executor, Callable[[QPROGRAM], MeasurementResult]],
     *,
     inverse_confusion_matrix: npt.NDArray[np.float64],
-) -> Callable[[QPROGRAM], float]:
+) -> Union[Executor, Callable[[QPROGRAM], MeasurementResult]]:
     """Returns a modified version of the input 'executor' which is
     error-mitigated with readout confusion inversion (RCI).
 
     Args:
         executor: A Mitiq executor that executes a circuit and returns the
             unmitigated ``MeasurementResult``.
-        observable: Observable to compute the expectation value of (required).
         inverse_confusion_matrix: The inverse confusion matrix to apply to the
             probability vector estimated with noisy measurement results.
 
     Returns:
         The error-mitigated version of the input executor.
     """
+    if not isinstance(executor, Executor):
+        executor_obj = Executor(executor)
```

Comment:
I haven't understood this pattern too well. In the case where an executor is not an `Executor`, then it must be a `Callable[[QPROGRAM], MeasurementResult]`, but what is the distinction here between these two? Can you give examples?

Reply:
`Executor` is a class that Mitiq uses to wrap executor functions. `Executor` is what Mitiq always uses under the hood to execute circuits.
However, in order to make the user's experience simpler and more direct, we also accept as a valid input a simple Python function instead of an `Executor` object. In this case, we internally wrap the function into an `Executor` object. 

I added some docs and comments in the next commit to help developers and maintainers. But please, if things are not clear, let me know.